### PR TITLE
Fix SIL optimizer bug with not validated decls

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

File: lib/SILOptimizer/Analysis/ClassHierarchyAnalysis.cpp [hunk 1]
```diff
@@ -32,7 +32,7 @@ class NominalTypeWalker: public ASTWalker {
 
   bool walkToDeclPre(Decl *D) override {
     auto *NTD = dyn_cast<NominalTypeDecl>(D);
-    if (!NTD)
+    if (!NTD || !NTD->hasInterfaceType())
```
Comment (swiftix) [inline lib/SILOptimizer/Analysis/ClassHierarchyAnalysis.cpp]: @slavapestov If `!NTD->hasInterfaceType()` is true, does it stay like this for the whole duration of compilation or may it become false at some point? I.e. may this NTD become a class that we should have taken into account?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

File: lib/SILOptimizer/Analysis/ClassHierarchyAnalysis.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/SILOptimizer/Analysis/ClassHierarchyAnalysis.cpp]: By the time the SIL optimizer runs, the type checker has been torn down so nothing "new" can get type checked that wasn't type checked before.

Comment (swiftix) [inline lib/SILOptimizer/Analysis/ClassHierarchyAnalysis.cpp]: OK. Than it should be fine.

**Review by swiftix — APPROVED** [review]

**Comment by slavapestov** [comment]
@aciidb0mb3r Does the swiftpm failure look familiar?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**